### PR TITLE
Add ability to use `workspaceFolder` in Visual studio and VSCode to interpolate settings

### LINKED ISSUE #889: Visual Studio extension various problems
I've found a few issues with the VS extension. Two of them are easy fixes I think, but I've never done VS extension development and I'm not sure what testing is required, so I'm hesitant to contribute a PR, and am hoping someone already set up can just make them quickly.

First, the README.md is wrong. At https://github.com/microsoft/cadl/blob/b162c3ca5b32e984d358fadf01cc8f63b2ab4a78/packages/cadl-vs/README.md, it says to use the configuration:

```json
{
  "cadl.cadl-server-path": "./mynestedproject/node_modules/@cadl-lang/compiler"
}
```

But that's wrong, it's `cadl.cadl-server.path`, not `-path`

Second, there's an issue here: https://github.com/microsoft/cadl/blob/main/packages/cadl-vs/src/VSExtension.cs#L87

If you hit this line (which happens if you hit https://github.com/microsoft/cadl/blob/main/packages/cadl-vs/src/VSExtension.cs#L225), then it fails:

```
Task Failed - Activating language client Cadl: Failed to activate Cadl language server!
File issue at https://github.com/microsoft/cadl

System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.ThrowHelper.ThrowKeyNotFoundException()
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Microsoft.Cadl.VisualStudio.LanguageClient.<ActivateAsync>d__20.MoveNext() in D:\a\1\s\packages\cadl-vs\src\VSExtension.cs:line 87
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.<InitializeAsync>d__117.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.TplExtensions.<WithTimeout>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.<TrackTaskAsync>d__118.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.LanguageServer.Client.RemoteLanguageClientInstance.<ActivateLanguageClientAsync>d__127.MoveNext()
```

I'm pretty sure the fix is just to make that line:

```
info.Environment[entry.Key] = entry.Value;
```

because `entry` is already an entry in `env`.

Third, it doesn't seem to work with relative paths. When I tried:

```json
{
  "cadl.cadl-server.path": ".\\src\\cadl\\cadl\\node_modules\\.bin\\cadl-server.cmd"
}
```

I hit the above error (suggesting it hadn't found the file, per https://github.com/microsoft/cadl/blob/main/packages/cadl-vs/src/VSExtension.cs#L214-L218), but when I replaced it with:

```json
{
  "cadl.cadl-server.path": "C:\\<path-to-my-repo>\\src\\cadl\\cadl\\node_modules\\.bin\\cadl-server.cmd"
}
```
It worked.

Note that the 2nd issue here (the bug in the env processing) means I had to specify the `.cmd` file explicitly, because if I tried to just follow the README.md and specify `node_modules/@cadl-lang/compiler`, then I hit this branch: https://github.com/microsoft/cadl/blob/main/packages/cadl-vs/src/VSExtension.cs#L221, which makes me hit https://github.com/microsoft/cadl/blob/main/packages/cadl-vs/src/VSExtension.cs#L225, which in turn makes me hit https://github.com/microsoft/cadl/blob/main/packages/cadl-vs/src/VSExtension.cs#L87, which is where the bug is

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #889

Changes: 
- Fix VS documentation using wrong setting name.
- Fix bug in VS with environment variable
- Add support for `${workspaceFolder}` interpolation in vs config
- Add support for `${workspaceFolder}` interpolation in vscode settings along side existing `workspaceRoot` that is deprecated.

### CODE REVIEW

File: packages/cadl-vs/src/VSExtension.cs [hunk 1]
```diff
@@ -195,12 +195,12 @@ private static string GetDevelopmentCadlServerPath()
...
-                var options = Environment.GetEnvironmentVariable("CADL_SERVER_NODE_OPTIONS");
-                var module = GetDevelopmentCadlServerPath();
-                return ("node.exe", new string[] { module, options }.Concat(args).ToArray(), env);
-            }
+            //if (InDevelopmentMode())
+            //{
+            //    var options = Environment.GetEnvironmentVariable("CADL_SERVER_NODE_OPTIONS");
+            //    var module = GetDevelopmentCadlServerPath();
+            //    return ("node.exe", new string[] { module, options }.Concat(args).ToArray(), env);
+            //}
```
Comment (nguerrera) [inline packages/cadl-vs/src/VSExtension.cs:203]: Is this meant to be commented out now or was this just for testing?

Comment (timotheeguerin) [reply packages/cadl-vs/src/VSExtension.cs:203]: testing :) I'll revert

File: packages/cadl-vs/src/VSExtension.cs [hunk 2]
```diff
@@ -84,9 +84,9 @@ public LanguageClient([Import] IVsFolderWorkspaceService workspaceService)
 
             foreach (var entry in env)
             {
-                info.Environment[entry.Key] = env[entry.Value];
+                info.Environment[entry.Key] = entry.Value;
             }
-#if env
+#if DEBUG
```
Comment (nguerrera) [inline packages/cadl-vs/src/VSExtension.cs:89]: Yikes

File: packages/cadl-vscode/README.md [hunk 3]
```diff
@@ -5,3 +5,22 @@ This provides provides [Cadl](https://github.com/microsoft/cadl) language suppor
...
+
+Cadl wil interpolate a few variables using this pattern `${<name>}`. For example `${workspaceFolder}`.
+
+Available variables:
+
+- `workspaceFolder`: Corespond to the root of your Visual Studio workspace.
+
+### `cadl.cadl-server.path`: Configure the server path
+
+There is cases where the Cadl project is located in a subfolder. In this case Cadl extension is not able to find the cadl compiler automatically and needs a little guidance.
```
Comment (nguerrera) [inline packages/cadl-vscode/README.md:19]:
```suggestion
There are cases where the Cadl project is located in a subfolder. In such cases, the Cadl extension is not able to find the cadl compiler automatically and needs a little guidance.
```

File: packages/cadl-vscode/README.md [hunk 4]
```diff
@@ -5,3 +5,22 @@ This provides provides [Cadl](https://github.com/microsoft/cadl) language suppor
...
+This setting provide the ability to configure where the cadl compiler is located.
```
Comment (nguerrera) [inline packages/cadl-vscode/README.md:20]:
```suggestion
This setting provides the ability to configure where the cadl compiler is located.
```

File: packages/cadl-vscode/src/extension.ts [hunk 5]
```diff
@@ -139,3 +141,24 @@ async function isFile(path: string) {
 export async function deactivate() {
   await client?.stop();
 }
+
+/**
+ * Resolve some of the VSCode variable.
```
Comment (nguerrera) [inline packages/cadl-vscode/src/extension.ts:146]:
```suggestion
 * Resolve some of the VSCode variables.
```

File: packages/cadl-vscode/src/extension.ts [hunk 6]
```diff
@@ -103,11 +103,13 @@ async function resolveCadlServer(context: ExtensionContext): Promise<Executable>
     const executable = process.platform === "win32" ? "cadl-server.cmd" : "cadl-server";
     return { command: executable, args, options };
   }
+  const workspaceFolder = workspace.workspaceFolders?.[0]?.uri?.fsPath ?? "";
+  const variableResolver = new VSCodeVariableResolver({
+    workspaceFolder,
+    workspaceRoot: workspaceFolder, // workspaceRoot is deprecated but we keeping it for now for legacy reason.
```
Comment (nguerrera) [inline packages/cadl-vscode/src/extension.ts:109]:
```suggestion
    workspaceRoot: workspaceFolder, // workspaceRoot is deprecated but we still support it for backwards compatibility.
```

**Review by nguerrera — APPROVED** [review]
Looks good apart from reverting the commented out code and some very small nits about comments and doc wording.

**Pull request merged**